### PR TITLE
add key manager ID check

### PR DESCRIPTION
**Pull request by matevz**
PR for #2459:
- adds a check whether key manager ID of the newly registered runtime points to a valid key manager in the registry.

### CODE REVIEW

File: .changelog/2459.feature.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+When registering a new runtime, check whether the given key manager ID points
```
Comment (kostko) [inline .changelog/2459.feature.md]: This is actually `breaking` as the same transaction can be processed differently by different versions.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 2]
```diff
@@ -402,6 +402,23 @@ func (app *registryApplication) registerRuntime(
...
 	}
 
+	if rt.Kind == registry.KindCompute {
+		// Find key manager runtime, if it is set.
+		var km *registry.Runtime
+		if rt.KeyManager != nil {
+			km, err = state.ImmutableState.Runtime(*rt.KeyManager)
+		}
+
+		if err != nil {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go]: This check should be in the upper if, e.g.:
```go
if km, err = state...; err != nil { ...
```

File: go/registry/api/api.go [hunk 3]
```diff
@@ -930,6 +930,34 @@ func VerifyRegisterRuntimeArgs(logger *logging.Logger, sigRt *SignedRuntime, isG
 	return &rt, nil
 }
 
+// VerifyRegisterComputeRuntimeArgs verifies ComputeRuntime-specific arguments for RegisterRuntime.
+func VerifyRegisterComputeRuntimeArgs(logger *logging.Logger, rt *Runtime, km *Runtime) error {
+	// Check runtime's key manager, if key manager ID is set.
+	if rt.KeyManager != nil {
+		// key manager runtime must exist.
```
Comment (kostko) [inline go/registry/api/api.go]:
```suggestion
		// Key manager runtime must exist.
```

File: go/registry/api/api.go [hunk 4]
```diff
@@ -930,6 +930,34 @@ func VerifyRegisterRuntimeArgs(logger *logging.Logger, sigRt *SignedRuntime, isG
...
+		if km == nil {
```
Comment (kostko) [inline go/registry/api/api.go]: This seems awkward, as mentioned in my previous review, it would be better to handle this via a `RuntimeLookup` interface, similar to the [existing `NodeLookup` interface](https://github.com/oasislabs/oasis-core/blob/master/go/registry/api/api.go#L274-L282). I will also need this interface in #2504.

So instead of passing `km`, pass the `RuntimeLookup` interface and call `runtimeLookup.Runtime(*rt.KeyManager)` here.

Comment (matevz) [reply go/registry/api/api.go]: I added `RuntimeLookup` with only single method `Runtime()` now. This method was already implemented before by `ImmutableState`.

File: go/registry/api/api.go [hunk 5]
```diff
@@ -930,6 +937,43 @@ func VerifyRegisterRuntimeArgs(logger *logging.Logger, sigRt *SignedRuntime, isG
...
+		if err != nil {
+			logger.Error("RegisterRuntime: error when fetching the runtime's key manager from registry",
+				"runtime", rt.ID,
+				"key_manager", rt.KeyManager,
+			)
+			return err
+		}
+
+		// Key manager runtime must exist.
+		if km == nil {
```
Comment (kostko) [inline go/registry/api/api.go]: I think that `Runtime()` should return `ErrNoSuchRuntime` if the runtime doesn't exist, so this additional check is not needed?

File: go/scheduler/tests/tester.go [hunk 6]
```diff
@@ -31,7 +31,7 @@ func SchedulerImplementationTests(t *testing.T, name string, backend api.Backend
 	rt, err := registryTests.NewTestRuntime(seed, nil)
 	require.NoError(err, "NewTestRuntime")
 
-	// Populate the registry with an entity and nodes.
+	// Populate the registryAPI with an entity and nodes.
```
Comment (kostko) [inline go/scheduler/tests/tester.go]: Too much search and replace?

File: go/scheduler/tests/tester.go [hunk 7]
```diff
@@ -126,7 +126,7 @@ func SchedulerImplementationTests(t *testing.T, name string, backend api.Backend
 
 	ensureValidCommittees(3, 1, int(rt.Runtime.TxnScheduler.GroupSize))
 
-	// Cleanup the registry.
+	// Cleanup the registryAPI.
```
Comment (kostko) [inline go/scheduler/tests/tester.go]: Too much search and replace?

File: go/registry/api/api.go [hunk 8]
```diff
@@ -930,6 +938,34 @@ func VerifyRegisterRuntimeArgs(logger *logging.Logger, sigRt *SignedRuntime, isG
 	return &rt, nil
 }
 
+// VerifyRegisterComputeRuntimeArgs verifies ComputeRuntime-specific arguments for RegisterRuntime.
```
Comment (kostko) [inline go/registry/api/api.go]:
```suggestion
// VerifyRegisterComputeRuntimeArgs verifies compute runtime-specific arguments for RegisterRuntime.
```

File: go/scheduler/tests/tester.go [hunk 9]
```diff
@@ -14,7 +14,7 @@ import (
 	consensusAPI "github.com/oasislabs/oasis-core/go/consensus/api"
 	epochtime "github.com/oasislabs/oasis-core/go/epochtime/api"
 	epochtimeTests "github.com/oasislabs/oasis-core/go/epochtime/tests"
-	registry "github.com/oasislabs/oasis-core/go/registry/api"
+	registryAPI "github.com/oasislabs/oasis-core/go/registry/api"
```
Comment (kostko) [inline go/scheduler/tests/tester.go]: Why this change? There doesn't seem to be any other changes in this file?
